body_wrapper.go

```diff
@@ -115,28 +133,47 @@ func (bw *bodyWrapper) GetBody() (io.ReadCloser, error) {
 		return nil, bw.readErr
 	}
 
-	// Lazy initialization
-	if !bw.isInitialized {
-		if initErr := bw.initialize(); initErr != nil {
-			return nil, initErr
+	if bw.isStoringInMemDisabled {
+		return nil, errors.New("body caching is disabled, cannot get body contents")
+	}
+
+	if !bw.isFullyRead {
+		if err := bw.readFull(); err != nil {
+			return nil, err
 		}
 	}
 
 	return ioutil.NopCloser(bytes.NewReader(bw.origBytes)), nil
 }
 
-func (bw *bodyWrapper) initialize() error {
-	if !bw.isInitialized {
-		bw.isInitialized = true
+// Disables storing body contents in memory and clears the cache
+func (bw *bodyWrapper) DisableBodyCaching() error {
+	bw.mu.Lock()
+	defer bw.mu.Unlock()
 
-		if bw.origReader != nil {
-			bw.origBytes, bw.readErr = ioutil.ReadAll(bw.origReader)
+	bw.isStoringInMemDisabled = true
+	bw.origBytes = nil
 
-			_ = bw.closeAndCancel()
-		}
-	}
+	return nil
+}
 
-	return bw.readErr
+// Reads the body fully, then cancels and closes the reader
+func (bw *bodyWrapper) readFull() error {
+	if bw.isFullyRead {
+		return errors.New("body is already fully read")
+	}
+	remainingBytes, err := ioutil.ReadAll(bw.origReader)
+	if err != nil {
+		bw.readErr = err
+		bw.isFullyRead = true // Prevent further reads
+		_ = bw.closeAndCancel()
+		return err
+	}
+	initialBytesLen := len(bw.origBytes)
+	bw.origBytes = append(bw.origBytes, remainingBytes...)
+	bw.isFullyRead = true
+	bw.currReader = bytes.NewReader(bw.origBytes[initialBytesLen:])
```

Comment:
Do we really need to set currReader in readFull? It seems that in places when readFull is invoked, the calling code is setting currReader by itself.